baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/BadAssert.java

```diff
@@ -0,0 +1,86 @@
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.CompileTimeConstantExpressionMatcher;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.AssertTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.tools.javac.code.Type;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "BadAssert",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = LinkType.CUSTOM,
+        severity = SeverityLevel.ERROR,
+        summary = "Assert statements should not be used. Asserts are generally enabled in tests but disabled in "
+                + "production, which means code can execute in production that is impossible to test.\n"
+                + "If you're confident that an 'assert' is required, this check may be suppressed using "
+                + "@SuppressWarnings(\"BadAssert\")")
+public final class BadAssert extends BugChecker implements BugChecker.AssertTreeMatcher {
+
+    private static final String LOGSAFE_PRECONDITIONS = "com.palantir.logsafe.Preconditions";
+    private final Matcher<ExpressionTree> compileTimeConstExpressionMatcher =
+            new CompileTimeConstantExpressionMatcher();
+
+    @Override
+    public Description matchAssert(AssertTree tree, VisitorState state) {
+        ExpressionTree condition = tree.getCondition();
+        ExpressionTree detail = tree.getDetail();
+        SuggestedFix.Builder fix = SuggestedFix.builder();
+        if (detail == null) {
+            fix.replace(
+                    tree,
+                    String.format(
+                            "if (!%s) { throw new IllegalStateException(); }", state.getSourceForNode(condition)));
+        } else if (isString(ASTHelpers.getType(detail), state)
+                && compileTimeConstExpressionMatcher.matches(detail, state)) {
+            fix.replace(
+                    tree,
+                    String.format(
+                            "%s.checkState(%s, %s)",
+                            SuggestedFixes.qualifyType(state, fix, LOGSAFE_PRECONDITIONS),
+                            state.getSourceForNode(condition),
+                            state.getSourceForNode(detail)));
+        } else {
+            String message = isString(ASTHelpers.getType(detail), state)
+                    ? state.getSourceForNode(detail)
+                    : "String.valueOf(" + state.getSourceForNode(detail) + ")";
+            fix.replace(
+                    tree,
+                    String.format(
+                            "if (!%s) { throw new IllegalStateException(%s); }",
```

Comment:
These can result in funny code when the input is `assert !value` -> `if (!!value) { throw...`. Not worth optimizing here imo because we have other tools to catch and resolve those, and the fixes are only for documentation, not enabled by default.